File: numberofboomerangs.py
class Solution(object):
    def numberOfBoomerangs(self, points):
        """
        :type points: List[List[int]]
        :rtype: int
        """
        import collections
        def distance(p1, p2):
            return (p1[0]-p2[0])**2 + (p1[1]-p2[1])**2
        #dic = collections.Counter()
        dic = {}
        n = len(points)
        counter = 0
        for i in xrange(n):
            for j in xrange(n):
                if i == j:
                    continue
                d = distance(points[i], points[j])
                #dic[d] += 1
                dic[d] = 1 + dic.get(d, 0)
            for v in dic.values():
                counter += v*(v-1)
            dic.clear()
                
        return counter
        

File: numberofboomerangs.cpp
class Solution {
public:
    int numberOfBoomerangs(vector<pair<int, int>>& points) {
        int cnt = 0;
        unordered_map<int, int> count;
        for (int i = 0; i < points.size(); i++) {
            for (int j = 0; j < points.size(); j++) {
                if (i == j) continue;
                count[getDistance(points[i], points[j])]++;
            }
            for (auto &e: count) {
                cnt += e.second*(e.second-1);
            }
            count.clear();
        }
        return cnt;
    }
    int getDistance(pair<int, int> &a, pair<int, int> &b) {
        return pow(abs(a.first-b.first), 2) + pow(abs(a.second-b.second), 2);
    }
};
